Bottom: c6abd6fdf5c0d57aebfe1b3ee3b17a455473716f
Top:    3cdbab1e9fb70e48c57020879cad450d4965b263
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-05-14 00:46:27 +0100

GTK/GL: Add facility to set an alpha multiplier for the current rendering


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 2a50cd1..a93c607 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -46,6 +46,7 @@ typedef struct render_priv {
   bool in_context;
   int subcomposite_stencil_bit;
   char *current_colorname;
+  double current_alpha_mult;
 } render_priv;
 
 
@@ -54,6 +55,7 @@ typedef struct hid_gc_struct
   HID *me_pointer;
 
   const char *colorname;
+  double alpha_mult;
   gint width;
   gint cap, join;
   gchar xor;
@@ -172,6 +174,7 @@ ghid_make_gc (void)
   rv = g_new0 (hid_gc_struct, 1);
   rv->me_pointer = &ghid_hid;
   rv->colorname = Settings.BackgroundColor;
+  rv->alpha_mult = 1.0;
   return rv;
 }
 
@@ -356,11 +359,13 @@ set_gl_color_for_gc (hidGC gc)
   double r, g, b, a;
 
   if (priv->current_colorname != NULL &&
-      strcmp (priv->current_colorname, gc->colorname) == 0)
+      strcmp (priv->current_colorname, gc->colorname) == 0 &&
+      priv->current_alpha_mult == gc->alpha_mult)
     return;
 
   free (priv->current_colorname);
   priv->current_colorname = strdup (gc->colorname);
+  priv->current_alpha_mult = gc->alpha_mult;
 
   if (gport->colormap == NULL)
     gport->colormap = gtk_widget_get_colormap (gport->top_window);
@@ -422,6 +427,7 @@ set_gl_color_for_gc (hidGC gc)
     }
   if (1) {
     double maxi, mult;
+    a *= gc->alpha_mult;
     if (!priv->trans_lines)
       a = 1.0;
     maxi = r;
@@ -450,6 +456,13 @@ ghid_set_color (hidGC gc, const char *name)
 }
 
 void
+ghid_set_alpha_mult (hidGC gc, double alpha_mult)
+{
+  gc->alpha_mult = alpha_mult;
+  set_gl_color_for_gc (gc);
+}
+
+void
 ghid_set_line_cap (hidGC gc, EndCapStyle style)
 {
   gc->cap = style;
